<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Systems Immunology Workshops - 19&nbsp; Tidyverse</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../session3/pSet3.html" rel="next">
<link href="../session3/count_data1.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Tidyverse</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Systems Immunology Workshops</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">Systems Immunology Workshops</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../preWork/pre_work.html" class="sidebar-item-text sidebar-link">Pre-Work</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../preWork/install.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Installing R and RStudio</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../preWork/rstudio.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to RStudio</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../session1/session1.html" class="sidebar-item-text sidebar-link">Session 1</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session1/types.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">R Syntax and Data Structures</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session1/prob.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Probability Primer</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session1/hypo_tests.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Distributions to Hypothesis Tests</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session1/factors.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Categorical Data in R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session1/choosing_tests.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Performing and choosing hypothesis tests</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session1/pSet1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Problem Set 1</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../session2/session2.html" class="sidebar-item-text sidebar-link">Session 2</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session2/packages.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Packages and Libraries</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session2/loading_functions.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Reading data into R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session2/pvals.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">P Values and Multiple Hypotheses</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session2/mult.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multiple Hypothesis Correction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session2/making_functions.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Functions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session2/session2problems.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Practice Exercises</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session2/pSet2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Problem Set 2</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../session3/session3.html" class="sidebar-item-text sidebar-link">Session 3</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session3/wrangling.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Data Wrangling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session3/matching_reordering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Matching and Reordering Data in R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session3/count_data1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Count Data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session3/tidyverse.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Tidyverse</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../session3/pSet3.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Problem Set 3</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#data-wrangling-with-tidyverse" id="toc-data-wrangling-with-tidyverse" class="nav-link active" data-scroll-target="#data-wrangling-with-tidyverse"><span class="toc-section-number">20</span>  Data Wrangling with Tidyverse</a>
  <ul class="collapse">
  <li><a href="#tidyverse-basics" id="toc-tidyverse-basics" class="nav-link" data-scroll-target="#tidyverse-basics"><span class="toc-section-number">20.1</span>  Tidyverse basics</a>
  <ul class="collapse">
  <li><a href="#pipes" id="toc-pipes" class="nav-link" data-scroll-target="#pipes"><span class="toc-section-number">20.1.1</span>  Pipes</a></li>
  <li><a href="#tibbles" id="toc-tibbles" class="nav-link" data-scroll-target="#tibbles"><span class="toc-section-number">20.1.2</span>  Tibbles</a></li>
  </ul></li>
  <li><a href="#experimental-data" id="toc-experimental-data" class="nav-link" data-scroll-target="#experimental-data"><span class="toc-section-number">20.2</span>  Experimental data</a></li>
  <li><a href="#analysis-goal-and-workflow" id="toc-analysis-goal-and-workflow" class="nav-link" data-scroll-target="#analysis-goal-and-workflow"><span class="toc-section-number">20.3</span>  Analysis goal and workflow</a></li>
  <li><a href="#instructions" id="toc-instructions" class="nav-link" data-scroll-target="#instructions"><span class="toc-section-number">20.4</span>  Instructions</a></li>
  <li><a href="#tidyverse-tools" id="toc-tidyverse-tools" class="nav-link" data-scroll-target="#tidyverse-tools"><span class="toc-section-number">20.5</span>  Tidyverse tools</a></li>
  <li><a href="#read-in-the-functional-analysis-results" id="toc-read-in-the-functional-analysis-results" class="nav-link" data-scroll-target="#read-in-the-functional-analysis-results"><span class="toc-section-number">20.6</span>  1. Read in the functional analysis results</a></li>
  <li><a href="#extract-only-the-go-biological-processes-bp-of-interest" id="toc-extract-only-the-go-biological-processes-bp-of-interest" class="nav-link" data-scroll-target="#extract-only-the-go-biological-processes-bp-of-interest"><span class="toc-section-number">20.7</span>  2. Extract only the GO biological processes (BP) of interest</a></li>
  <li><a href="#select-only-the-columns-needed-for-visualization" id="toc-select-only-the-columns-needed-for-visualization" class="nav-link" data-scroll-target="#select-only-the-columns-needed-for-visualization"><span class="toc-section-number">20.8</span>  3. Select only the columns needed for visualization</a></li>
  <li><a href="#order-go-processes-by-significance-adjusted-p-values" id="toc-order-go-processes-by-significance-adjusted-p-values" class="nav-link" data-scroll-target="#order-go-processes-by-significance-adjusted-p-values"><span class="toc-section-number">20.9</span>  4. Order GO processes by significance (adjusted p-values)</a></li>
  <li><a href="#rename-columns-to-be-more-intuitive" id="toc-rename-columns-to-be-more-intuitive" class="nav-link" data-scroll-target="#rename-columns-to-be-more-intuitive"><span class="toc-section-number">20.10</span>  5. Rename columns to be more intuitive</a></li>
  <li><a href="#create-additional-metrics-for-plotting-e.g.-gene-ratios" id="toc-create-additional-metrics-for-plotting-e.g.-gene-ratios" class="nav-link" data-scroll-target="#create-additional-metrics-for-plotting-e.g.-gene-ratios"><span class="toc-section-number">20.11</span>  6. Create additional metrics for plotting (e.g.&nbsp;gene ratios)</a></li>
  <li><a href="#compare-code" id="toc-compare-code" class="nav-link" data-scroll-target="#compare-code"><span class="toc-section-number">20.12</span>  Compare code</a>
  <ul class="collapse">
  <li><a href="#additional-resources" id="toc-additional-resources" class="nav-link" data-scroll-target="#additional-resources"><span class="toc-section-number">20.12.1</span>  Additional resources</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Tidyverse</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="data-wrangling-with-tidyverse" class="level1" data-number="20">
<h1 data-number="20"><span class="header-section-number">20</span> Data Wrangling with Tidyverse</h1>
<p>The <a href="https://www.tidyverse.org/packages/">Tidyverse suite of integrated packages</a> are designed to work together to make common data science operations more user friendly. The packages have functions for data wrangling, tidying, reading/writing, parsing, and visualizing, among others. There is a freely available book, <a href="http://r4ds.had.co.nz/index.html">R for Data Science</a>, with detailed descriptions and practical examples of the tools available and how they work together. We will explore the basic syntax for working with these packages, as well as, specific functions for data wrangling with the ‘dplyr’ package and data visualization with the ‘ggplot2’ package.</p>
<p><img src="../img/tidyverse_website.png" class="img-fluid"></p>
<section id="tidyverse-basics" class="level2" data-number="20.1">
<h2 data-number="20.1" class="anchored" data-anchor-id="tidyverse-basics"><span class="header-section-number">20.1</span> Tidyverse basics</h2>
<p>The Tidyverse suite of packages introduces users to a set of data structures, functions and operators to make working with data more intuitive, but is slightly different from the way we do things in base R. <strong>Two important new concepts we will focus on are pipes and tibbles</strong>.</p>
<p>Before we get started with pipes or tibbles, let’s load the library:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="pipes" class="level3" data-number="20.1.1">
<h3 data-number="20.1.1" class="anchored" data-anchor-id="pipes"><span class="header-section-number">20.1.1</span> Pipes</h3>
<p>Stringing together commands in R can be quite daunting. Also, trying to understand code that has many nested functions can be confusing.</p>
<p>To make R code more human readable, the Tidyverse tools use the pipe, <code>%&gt;%</code>, which was acquired from the <code>magrittr</code> package and is now part of the <code>dplyr</code> package that is installed automatically with Tidyverse. <strong>The pipe allows the output of a previous command to be used as input to another command instead of using nested functions.</strong></p>
<blockquote class="blockquote">
<p><strong>NOTE:</strong> Shortcut to write the pipe is <kbd>shift</kbd> + <kbd>command</kbd> + <kbd>M</kbd></p>
</blockquote>
<p>An example of using the pipe to run multiple commands:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">## A single command</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="dv">83</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 9.110434</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Base R method of running more than one command</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">sqrt</span>(<span class="dv">83</span>), <span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 9.11</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Running more than one command with piping</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="dv">83</span>) <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 9.11</code></pre>
</div>
</div>
<p>The pipe represents a much easier way of writing and deciphering R code, and so we will be taking advantage of it, when possible, as we work through the remaining lesson.</p>
</section>
<section id="tibbles" class="level3" data-number="20.1.2">
<h3 data-number="20.1.2" class="anchored" data-anchor-id="tibbles"><span class="header-section-number">20.1.2</span> Tibbles</h3>
<p>A core component of the <a href="http://tidyverse.org/">tidyverse</a> is the <a href="http://tibble.tidyverse.org/">tibble</a>. <strong>Tibbles are a modern rework of the standard <code>data.frame</code>, with some internal improvements</strong> to make code more reliable. They are data frames, but do not follow all of the same rules. For example, tibbles can have numbers/symbols for column names, which is not normally allowed in base R.</p>
<p><strong>Important: <a href="http://tidyverse.org/">tidyverse</a> is very opininated about row names</strong>. These packages insist that all column data (e.g.&nbsp;<code>data.frame</code>) be treated equally, and that special designation of a column as <code>rownames</code> should be deprecated. <a href="http://tibble.tidyverse.org/">Tibble</a> provides simple utility functions to handle rownames: <code>rownames_to_column()</code> and <code>column_to_rownames()</code>.</p>
<p>Tibbles can be created directly using the <code>tibble()</code> function or data frames can be converted into tibbles using <code>as_tibble(name_of_df)</code>.</p>
<blockquote class="blockquote">
<p><strong>NOTE:</strong> The function <code>as_tibble()</code> will ignore row names, so if a column representing the row names is needed, then the function <code>rownames_to_column(name_of_df)</code> should be run prior to turning the data.frame into a tibble. Also, <code>as_tibble()</code> will not coerce character vectors to factors by default.</p>
</blockquote>
</section>
</section>
<section id="experimental-data" class="level2" data-number="20.2">
<h2 data-number="20.2" class="anchored" data-anchor-id="experimental-data"><span class="header-section-number">20.2</span> Experimental data</h2>
<p>We’re going to explore the Tidyverse suite of tools to wrangle our data to prepare it for visualization. Make sure you have the file called <code>gprofiler_results_Mov10oe.tsv</code>.</p>
<p><strong>The dataset:</strong></p>
<ul>
<li>Represents the <strong>functional analysis results</strong>, including the biological processes, functions, pathways, or conditions that are over-represented in a given list of genes.</li>
<li>Our gene list was generated by <strong>differential gene expression analysis</strong> and the genes represent differences between <strong>control mice</strong> and <strong>mice over-expressing a gene involved in RNA splicing</strong>.</li>
</ul>
<p>The functional analysis that we will focus on involves <strong>gene ontology (GO) terms</strong>, which:</p>
<ul>
<li>describe the roles of genes and gene products</li>
<li>organized into three controlled vocabularies/ontologies (domains):
<ul>
<li>biological processes (BP)</li>
<li>cellular components (CC)</li>
<li>molecular functions (MF)</li>
</ul></li>
</ul>
<p><img src="../img/mov10_FA.png" width="1200"></p>
</section>
<section id="analysis-goal-and-workflow" class="level2" data-number="20.3">
<h2 data-number="20.3" class="anchored" data-anchor-id="analysis-goal-and-workflow"><span class="header-section-number">20.3</span> Analysis goal and workflow</h2>
<p><strong>Goal:</strong> <em>Visually compare the most significant biological processes (BP) based on the number of associated differentially expressed genes (gene ratios) and significance values by creating the following plot:</em></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/dotplot6.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">dotplot6</figcaption><p></p>
</figure>
</div>
<p>To wrangle our data in preparation for the plotting, we are going to use the Tidyverse suite of tools to wrangle and visualize our data through several steps:</p>
<ol type="1">
<li>Read in the functional analysis results</li>
<li>Extract only the GO biological processes (BP) of interest</li>
<li>Select only the columns needed for visualization</li>
<li>Order by significance (p-adjusted values)</li>
<li>Rename columns to be more intuitive</li>
<li>Create additional metrics for plotting (e.g.&nbsp;gene ratios)</li>
<li>Plot results</li>
</ol>
</section>
<section id="instructions" class="level2" data-number="20.4">
<h2 data-number="20.4" class="anchored" data-anchor-id="instructions"><span class="header-section-number">20.4</span> Instructions</h2>
<p>Find a partner (or a group of 3 if needed). Choose one person to go through the following steps using Tidyverse, and the other using base R. It is recommended that the person with more experience attempt the steps in base R.</p>
</section>
<section id="tidyverse-tools" class="level2" data-number="20.5">
<h2 data-number="20.5" class="anchored" data-anchor-id="tidyverse-tools"><span class="header-section-number">20.5</span> Tidyverse tools</h2>
<p>While all of the tools in the Tidyverse suite are deserving of being explored in more depth, we are going to investigate more deeply the reading (<code>readr</code>), wrangling (<code>dplyr</code>), and plotting (<code>ggplot2</code>) tools.</p>
</section>
<section id="read-in-the-functional-analysis-results" class="level2" data-number="20.6">
<h2 data-number="20.6" class="anchored" data-anchor-id="read-in-the-functional-analysis-results"><span class="header-section-number">20.6</span> 1. Read in the functional analysis results</h2>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" aria-current="page">Tidyverse</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">Base R</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<p>While the base R packages have perfectly fine methods for reading in data, the <code>readr</code> and <code>readxl</code> Tidyverse packages offer additional methods for reading in data. Let’s read in our tab-delimited functional analysis results <code>gprofiler_results_Mov10oe.tsv</code> using <code>read_delim()</code>. Name the dataframe <code>functional_GO_results</code>.</p>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<p>Use one of the base R <code>read.X</code> functions to read in the tab delimited file <code>gprofiler_results_Mov10oe.tsv</code>. Name the dataframe <code>functional_GO_results</code>.</p>
</div>
</div>
</div>
<p>Double check the data types and format of your dataframe. Do the methods yield the same result? Convert anything you think should be a factor into a factor.</p>
<p><em>NOTE: A large number of tidyverse functions will work with both tibbles and dataframes, and the data structure of the output will be identical to the input. However, there are some functions that will return a tibble (without row names), whether or not a tibble or dataframe is provided.</em></p>
</section>
<section id="extract-only-the-go-biological-processes-bp-of-interest" class="level2" data-number="20.7">
<h2 data-number="20.7" class="anchored" data-anchor-id="extract-only-the-go-biological-processes-bp-of-interest"><span class="header-section-number">20.7</span> 2. Extract only the GO biological processes (BP) of interest</h2>
<p>Now that we have our data, we will need to wrangle it into a format ready for plotting. To extract the biological processes of interest, we only want those rows where the <code>domain</code> is equal to <code>BP</code>.</p>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true">Tidyverse</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-2" role="tab" aria-controls="tabset-2-2" aria-selected="false">Base R</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<p>For all of our data wrangling steps we will be using tools from the <a href="http://dplyr.tidyverse.org/">dplyr</a> package, which is a swiss-army knife for data wrangling of data frames.</p>
<p>To extract the biological processes of interest, we only want those rows where the <code>domain</code> is equal to <code>BP</code>, which we can do using the <code>filter()</code> function.</p>
<p>To filter rows of a data frame/tibble based on values in different columns, we give a logical expression as input to the <code>filter()</code> function to return those rows for which the expression is TRUE.</p>
<p>Perform an additional filtering step to only keep those rows where the <code>relative.depth</code> is greater than 4.</p>
</div>
<div id="tabset-2-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-2-tab">
<p>Use a conditional expression and indexing (<code>[]</code>) to extract the rows where the <code>domain</code> is equal to <code>BP</code>.</p>
<p>Perform an additional indexing step to only keep those rows where the <code>relative.depth</code> is greater than 4.</p>
</div>
</div>
</div>
<p>Now we have returned only those rows with a <code>domain</code> of <code>BP</code>. <strong>How have the dimensions of our results changed?</strong></p>
</section>
<section id="select-only-the-columns-needed-for-visualization" class="level2" data-number="20.8">
<h2 data-number="20.8" class="anchored" data-anchor-id="select-only-the-columns-needed-for-visualization"><span class="header-section-number">20.8</span> 3. Select only the columns needed for visualization</h2>
<p>For visualization purposes, we are only interested in the columns related to the GO terms, the significance of the terms, and information about the number of genes associated with the terms.</p>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-3-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-1" role="tab" aria-controls="tabset-3-1" aria-selected="true">Tidyverse</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-2" role="tab" aria-controls="tabset-3-2" aria-selected="false">Base R</a></li></ul>
<div class="tab-content">
<div id="tabset-3-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-3-1-tab">
<p>To extract columns from a data frame/tibble we can use the <code>select()</code> function. In contrast to base R, we do not need to put the column names in quotes for selection.</p>
<p>Select the columns <code>term.id, term.name, p.value, query.size, term.size, overlap.size, intersection</code>.</p>
</div>
<div id="tabset-3-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-2-tab">
<p>Index the columns<code>term.id, term.name, p.value, query.size, term.size, overlap.size, intersection</code>.</p>
</div>
</div>
</div>
<p>Both indesing and the <code>select()</code> function also allows for negative selection. However, <code>select</code> allows for negtaive selection using column names, while in base R we can only do so with indexes. Note that we need to put the column names inside of the combine (<code>c()</code>) function with a <code>-</code> preceding it for this functionality.</p>
<p>To use column names in base R, we have to use <code>%in%</code>:</p>
<pre><code># Selecting columns to keep
idx &lt;- !(colnames(functional_GO_results) %in% c("query.number", "significant", "recall", "precision", "subgraph.number", "relative.depth", "domain"))</code></pre>
</section>
<section id="order-go-processes-by-significance-adjusted-p-values" class="level2" data-number="20.9">
<h2 data-number="20.9" class="anchored" data-anchor-id="order-go-processes-by-significance-adjusted-p-values"><span class="header-section-number">20.9</span> 4. Order GO processes by significance (adjusted p-values)</h2>
<p>Now that we have only the rows and columns of interest, let’s arrange these by significance, which is denoted by the adjusted p-value.</p>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-4-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-1" role="tab" aria-controls="tabset-4-1" aria-selected="true">Tidyverse</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-2" role="tab" aria-controls="tabset-4-2" aria-selected="false">Base R</a></li></ul>
<div class="tab-content">
<div id="tabset-4-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-4-1-tab">
<p>Sort the rows by adjusted p-value with the <code>arrange()</code> function.</p>
</div>
<div id="tabset-4-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-2-tab">
<p>Sort the rows by adjusted p-value with the <code>order()</code> function.</p>
</div>
</div>
</div>
<p><em>NOTE: If you wanted to arrange in descending order, then you could have run the following instead:</em></p>
<pre><code># Order by adjusted p-value descending
functional_GO_results &lt;- functional_GO_results %&gt;%
arrange(desc(p.value))</code></pre>
<p><em>NOTE: Ordering variables in <code>ggplot2</code> is a bit different. <a href="https://www.r-graph-gallery.com/267-reorder-a-variable-in-ggplot2.html">This post</a> introduces a few ways of ordering variables in a plot.</em></p>
</section>
<section id="rename-columns-to-be-more-intuitive" class="level2" data-number="20.10">
<h2 data-number="20.10" class="anchored" data-anchor-id="rename-columns-to-be-more-intuitive"><span class="header-section-number">20.10</span> 5. Rename columns to be more intuitive</h2>
<p>While not necessary for our visualization, renaming columns more intuitively can help with our understanding of the data. Let’s rename the <code>term.id</code> and <code>term.name</code> columns.</p>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-5-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-1" role="tab" aria-controls="tabset-5-1" aria-selected="true">Tidyverse</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-5-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-2" role="tab" aria-controls="tabset-5-2" aria-selected="false">Base R</a></li></ul>
<div class="tab-content">
<div id="tabset-5-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-5-1-tab">
<p>Rename <code>term.id</code> and <code>term.name</code> to <code>GO_id</code> and <code>GO_term</code> using the <code>rename</code> function. Note that you may need to call rename as <code>dplyr::rename</code>, since rename is a common function name in other packages.</p>
<p>The syntax is <code>new_name = old_name</code>.</p>
</div>
<div id="tabset-5-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-5-2-tab">
<p>Rename <code>term.id</code> and <code>term.name</code> to <code>GO_id</code> and <code>GO_term</code> using <code>colnames</code> and indexing.</p>
</div>
</div>
</div>
</section>
<section id="create-additional-metrics-for-plotting-e.g.-gene-ratios" class="level2" data-number="20.11">
<h2 data-number="20.11" class="anchored" data-anchor-id="create-additional-metrics-for-plotting-e.g.-gene-ratios"><span class="header-section-number">20.11</span> 6. Create additional metrics for plotting (e.g.&nbsp;gene ratios)</h2>
<p>Finally, before we plot our data, we need to create a couple of additional metrics. Let’s generate gene ratios to reflect the number of DE genes associated with each GO process relative to the total number of DE genes.</p>
<p>This is calculated as <code>gene_ratio = overlap.size / query.size</code>.</p>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-6-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-1" role="tab" aria-controls="tabset-6-1" aria-selected="true">Tidyverse</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-6-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-2" role="tab" aria-controls="tabset-6-2" aria-selected="false">Base R</a></li></ul>
<div class="tab-content">
<div id="tabset-6-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-6-1-tab">
<p>The <code>mutate()</code> function enables you to create a new column from an existing column.</p>
</div>
<div id="tabset-6-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-6-2-tab">
<p>Create a new column in the dataframe using the <code>$</code> syntax or <code>cbind</code>.</p>
</div>
</div>
</div>
<p>The <code>mutate()</code> function enables you to create a new column from an existing column.</p>
<p><img src="../img/bp_oe_ready_to_plot.png" width="1200"></p>
</section>
<section id="compare-code" class="level2" data-number="20.12">
<h2 data-number="20.12" class="anchored" data-anchor-id="compare-code"><span class="header-section-number">20.12</span> Compare code</h2>
<p>Take a look at your code verses your partner’s code. Which method do you think results in cleaner, more readable code? Which steps were easier in base R, and which in Tidyverse?</p>
<section id="additional-resources" class="level3" data-number="20.12.1">
<h3 data-number="20.12.1" class="anchored" data-anchor-id="additional-resources"><span class="header-section-number">20.12.1</span> Additional resources</h3>
<ul>
<li><a href="http://r4ds.had.co.nz">R for Data Science</a></li>
<li><a href="http://varianceexplained.org/r/teach-tidyverse/">teach the tidyverse</a></li>
<li><a href="http://style.tidyverse.org/">tidy style guide</a></li>
</ul>
<hr>
<p><em>The materials in this lesson have been adapted from work created by the (HBC)](http://bioinformatics.sph.harvard.edu/) and Data Carpentry (http://datacarpentry.org/). These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.</em></p>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../session3/count_data1.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Count Data</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../session3/pSet3.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Problem Set 3</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>