# Problem Set 6

Please submit this problem set either as a R notebook (quarto or RMarkdown) or as an R script and write-up as needed.

## Problem 1: Quality Control

Take a closer look at the `qcstats` data frame computed in the quality control section above. Produce boxplots for (i) the total count for each cell (`sum`), (ii) the number of expressed features (`detected`), and (iii) the percentage of reads mapped to mitochondrial genes (`subsets_Mito_percent`). Identify outliers for all three QC criteria, ie those cells that are below the 5% quantile for `sum` and `detected`, and those cells that are above the 95% quantile for `subsets_Mito_percent` (Hint: see the `quantile` function).

## Problem 2: Dimensionality reduction
    
-   The choice of the number of PCs that we retain for downstream analysis should be based on how much of the total variance is explained by each PC. Therefore, visualize the percentage of variance explained for each PC in a scree plot and identify the elbow point, ie the point where the amount of variance explained seem to level off. Hint: inspect the `"percentVar"` attribute of the `reducedDim` slot using `attr`.
-   A key parameter of visualizing scRNA-seq data in a *t*-SNE plot is the `perplexity` parameter, which determines the granularity of the visualization. Produce *t*-SNE plots for different settings of the `perplexity` parameter (chose values between 5 and 100). How do increasing values of the `perplexity` parameter influence the visualization?
-   Important parameters for visualizing scRNA-seq data in a UMAP plot are the number of neighbors (`n_neighbors`) and the minimum distance between embedded points (`min_dist`). Explore the effect of these parameters on the granularity of the output for different settings of the `n_neighbors` parameter (chose integer values between 2 and 100) and different settings of the `min_dist` parameter (chose values between 0 and 1).
    
## Problem 3: Analysis

The [scRNAseq](https://bioconductor.org/packages/scRNAseq) package provides gene-level counts for a collection of public scRNA-seq datasets, stored as `SingleCellExperiment` objects with annotated cell- and gene-level metadata. Consult the vignette of the [scRNAseq](https://bioconductor.org/packages/scRNAseq) package to inspect all available datasets and select a dataset of your choice. Perform a typical scRNA-seq analysis on this dataset including QC, normalization, feature selection, dimensionality reduction, clustering, and marker gene detection.

